blockstack/gaia
# Https support

## Goal of PR

Implement support for the [Node.js https server](https://nodejs.org/docs/latest-v10.x/api/https.html) in order to simplify hub deployments. This makes it easier to orchestrate hub deployments for various compute providers and DNS/DDNS providers. 


## Implementation

Two options for enabling https are provided. See the updated `hub/config-schema.json` file diff for all supported config options. These are also all configurable via env vars. 

### Using TLS cert files

Supports both cert files in both the `PFX / PKCS12` format and the `PEM key & cert chain` format. 

Example of the config values for typical `pfx` cert file usage:
```json
{
  "enableHttps": "cert_files",
  "tlsCertConfig": {
    "pfxFile": "~/.config/ssl/cert.pfx"
  }
}
```

Example of the config values for the typical `PEM` files usage: 
```json
{
  "enableHttps": "cert_files",
  "tlsCertConfig": {
    "keyFile": "~/.config/ssl/key.pem",
    "certFile": "~/.config/ssl/cert.pem"
  }
}
```

### Using ACMEv2 client

This uses the [`greenlock-express`](https://www.npmjs.com/package/greenlock-express) middleware to provide support for Lets Encrypt v2 (i.e. [`ACME draft-12`](https://tools.ietf.org/html/draft-ietf-acme-acme-12)). 

The default ACME standard [challenge type `http-01`](https://letsencrypt.org/docs/challenge-types/#http-01-challenge) is used. This requires specifying the domain name(s) in the config _or_ ensuring that the server hostname is the intended domain name. 


However, the lib supports extensible ACME `challenge type` modules which a Gaia hub deployment system can use to easily add support for various DNS/DDNS services. Modules are already available for Google Cloud DNS, AWS (S3, Route53), Azure, CloudFlare, Digital Ocean, Namecheap, Godaddy, and more. 


Example of minimum config required to use (requires that server hostname is set to the intended domain name):
```json
{
  "enableHttps": "acme",
  "acmeConfig": {
    "email": "matt@example.com",
    "agreeTos": true
  }
}
```

Example of config where the domain name(s) are specified:
```json
{
  "enableHttps": "acme",
  "acmeConfig": {
    "email": "matt@example.com",
    "agreeTos": true,
    "approveDomains": ["gaia-testing.example.com", "gaia-testing2.example.com"]
  }
}
```


## Manual Testing

The ACME setup can be tested by setting up a Gaia hub on a VPS, pointing a domain to the server IP, and using the above config to enable `"enableHttps": "acme"`. The domain name can be specified in the hub config, or in the server's hostname config (e.g.: `hostname --fqdn` should show your domain name). 
I tested both methods using a Digital Ocean droplet and a namecheap subdomain A record. 


The TLS cert file setup can be tested by using the same resources from above, but with config set to `"enableHttps": "cert_files"`, and the `tlsCertConfig` options can be pointed to the cert PEM files that were generated by Lets Encrypt. 


## Automated Testing

The manual cert file config is covered in unit tests using self-signed localhost certs. 

The ACME setup is not covered in unit tests, except for the basic config value tests. The actual logic for using this lib is only a few lines of code -- essentially just:
```ts
// acmeClient.ts
import * as glx from 'greenlock-express'
export function createGlx(app: ExpressApp, acmeConfig: AcmeConfigInterface) {
  const opts: glx.GreenlockExpressOptions = { ... }
  return glx.create(opts)
}

// index.ts
const server = acme.createGlx(app, conf.acmeConfig)
server.listen(conf.port, conf.httpsPort, () => logHttpListen(), () => logHttpsListen())
```

The `greenlock-express` lib could be mocked to achieve code coverage, but it would be a useless test. This feature needs integration tests, which requires devops work for setting up a DNS API and likely a server, and it likely would not be possible to run automatically on commits due to rate limits. 


## Submission Checklist


      contains correct TypeFlow annotations.


      would be expected to interact with them.
